iree/compiler/Conversion/VectorToLLVM/VectorContractToAArch64InlineAsmOp.cpp

```diff
@@ -0,0 +1,174 @@
+/// performas a single matrix-vector product and to compute matmul(A, B) with
+/// matrix-vector products B is transposed.
+struct ConvertVectorContract4x4x4_i8i8i32_ToAArch64InlineAsmPattern
+    : public OpRewritePattern<vector::ContractionOp> {
+ public:
+  using OpRewritePattern<vector::ContractionOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::ContractionOp contractionOp,
+                                PatternRewriter &rewriter) const override {
+    auto lhsType = contractionOp.lhs().getType().cast<VectorType>();
+    auto rhsType = contractionOp.rhs().getType().cast<VectorType>();
+    auto accType = contractionOp.acc().getType().cast<VectorType>();
+    auto lhsShape = lhsType.getShape();
+    auto rhsShape = rhsType.getShape();
+    if (lhsShape[0] != 4 || lhsShape[1] != 4 || rhsShape[0] != 4 ||
+        rhsShape[1] != 4)
+      return failure();
+
+    Value inLhs = contractionOp.lhs();
+    Value inRhs = contractionOp.rhs();
+
+    if (accType.getElementType() != rewriter.getIntegerType(32)) {
+      return failure();
+    }
+    // TODO(ataei) : Remove this when linalg -> vector.contract
+    if (lhsType.getElementType() != rewriter.getIntegerType(8)) {
+      if (auto parentOp = cast<SignExtendIOp>(inLhs.getDefiningOp())) {
+        inLhs = parentOp.value();
+      } else {
+        return failure();
+      }
+    }
+    if (rhsType.getElementType() != rewriter.getIntegerType(8)) {
+      if (auto parentOp = cast<SignExtendIOp>(inRhs.getDefiningOp())) {
+        inRhs = parentOp.value();
+      } else {
+        return failure();
+      }
+    }
+    auto loc = contractionOp.getLoc();
+
+    auto dstVec =
+        llvm::to_vector<4>(llvm::map_range(llvm::seq<int>(0, 4), [&](int i) {
+          return rewriter.create<vector::ExtractOp>(loc, contractionOp.acc(),
+                                                    i);
+        }));
+
+    auto flattnedVectorType = VectorType::get({16}, rewriter.getIntegerType(8));
+
+    auto lhs =
+        rewriter.create<vector::ShapeCastOp>(loc, flattnedVectorType, inLhs);
+
+    auto inRhsTransposed = rewriter.create<vector::TransposeOp>(
+        loc, inRhs, ArrayRef<int64_t>({1, 0}));
+
+    auto rhs = rewriter.create<vector::ShapeCastOp>(loc, flattnedVectorType,
+                                                    inRhsTransposed);
+
+    auto int32x4VType = VectorType::get({4}, rewriter.getIntegerType(32));
+
+    auto returnType = LLVM::LLVMStructType::getLiteral(
+        rewriter.getContext(),
+        {int32x4VType, int32x4VType, int32x4VType, int32x4VType});
+
+    /// TODO(ataei): We an have a formatter like the c++ inline asm to
+    /// ssa-values to string names which will make the inline-assembly
+    /// statements more redable e.g :
+    /// sdot ${dstVec_0}.4s, ${lhs}.16b,${rhs}.4b[0]
+    auto packedResult = rewriter.create<LLVM::InlineAsmOp>(
+        loc, returnType,
+        ArrayRef<Value>({lhs, rhs, dstVec[0], dstVec[1], dstVec[2], dstVec[3]}),
```

Comment:
@nicolasvasilache Related to our conversation yesterday but down one level, the absent of 2D native LLVM `vector<4x4xi8>` constrained our IR to insert `shape_casts` everywhere. I was wondering as future ISA are going to work on 2D tiles and even few years old ISA like the bellow `sdot` have the concept of lanes in a vector it make sense to represent that in LLVM!

Reply:
First, glad to see this used, very nice :) !

Second, yes, Intel is the first to introduce a proper 2-D type for their AMX tiles, I expect this to generalize over time.